File: practicas/QytetetRuby/pkg/QytetetRuby-0.0.1/lib/tablero.rb
#encoding :utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "tipo_casilla.rb"
require_relative "casilla.rb"
require_relative "titulo_propiedad"

module ModeloQytetet
  class Tablero
    
    def initialize
      @casillas = Array.new
      @carcel = nil
      
      inicializar()
    end
    
    public
    attr_reader :casillas, :carcel
    
    def esCasillaCarcel(numeroCasilla)
      raise NotImplementedError 
    end
    
    def inicializar()  

      @casillas<< Casilla.casilla_sin_titulo(0, TipoCasilla::SALIDA)
      @casillas<< Casilla.new(1, TipoCasilla::CALLE,  TituloPropiedad.new("HOLA",0,0,0,0,0))
      @casillas<< Casilla.casilla_sin_titulo(2, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(3, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(4, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(5, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(6, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(7, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(8, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(10, TipoCasilla::CARCEL)
      @casillas<< Casilla.casilla_sin_titulo(11, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(12, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(13, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(14, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(15, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(16, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(17, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(18, TipoCasilla::SORPRESA)
      @casillas<< Casilla.casilla_sin_titulo(19, TipoCasilla::SORPRESA)

      @carcel = @casillas.select{|i| i.tipo == TipoCasilla::CARCEL}
    end
     private :inicializar
    
    protected
    def obtenerCasillaFinal(casilla, desplazamiento) 
      raise NotImplementedError 
    end
    
    def obtenerCasillaNumero(numeroCasilla)
      raise NotImplementedError 
    end
    
    public
    def to_s
      "Tablero: \n casillas:{ #{@casillas} } \n carcel: #{@carcel}"
    end
  end
end

File: QytetetJava/src/modeloqytetet/Tablero.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;
import java.util.ArrayList;

/**
 *
 * @author venrra
 */
public class Tablero {
    private ArrayList<Casilla> casillas;
    private Casilla carcel;
    
    Tablero() {
        this.inicializar(); 
    }
    
    boolean esCasillaCarcel(int numeroCasilla){
       return numeroCasilla == carcel.getNumeroCasilla();
    }
    
    Casilla getCarcel() {
        return carcel;
    }
    
    public ArrayList<Casilla> getCasillas() {
        return casillas;
    }
    
    // Se genera e inicializa el tablero con todas las casillas ordenadas del 0 al 19 intercalando los tipos
    private void inicializar(){
        casillas = new ArrayList<>();
        casillas.add(new Casilla(0,TipoCasilla.SALIDA));
        casillas.add(new Casilla(1, new TituloPropiedad("HOLA",0,0,0,0,0)));
        casillas.add(new Casilla(2, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(3, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(4, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(5, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(6, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(7, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(8, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(9, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(10,TipoCasilla.CARCEL));
        casillas.add(new Casilla(11, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(12, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(13, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(14, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(15, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(16, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(17, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(18, TipoCasilla.SORPRESA));
        casillas.add(new Casilla(19, TipoCasilla.SORPRESA));
       
        carcel = new Casilla(10, TipoCasilla.CARCEL);
    }
    
    Casilla obtenerCasillaFinal(Casilla casilla, int desplazamiento){
        int siguiente=0;
        siguiente = desplazamiento + casilla.getNumeroCasilla();
        siguiente %= casillas.size();
        
        return casillas.get(siguiente);
    }
    
    Casilla obtenerCasillaNumero(int numeroCasilla){
        /*
        boolean encontrado = false;
        Casilla aux;
        for (int i=0; i<casillas.size() && encontrado==false; i++ )
            if ( casillas.get(i).getNumeroCasilla() == numeroCasilla ){
                aux = casillas.get(i);
            }
        return aux;
        */
        return casillas.get(numeroCasilla);
    }
    
    @Override
    public String toString() {
        return "Tablero: \n" + "\nTodas las casillas=\n" + casillas + "\ncarcel=" + carcel + '}' + "\n";
    }
}
